### PR TITLE
Add an ability to except blocks' digging & hitting particles from multiplying color

### PR DESCRIPTION
**Pull request by Nullcaller**
I added an ability to except blocks' digging & hitting particles from multiplying color. To prevent particles from coloring, you need to implement IBlockColor from net.minecraft.renderer.color in your Block class, and to return true in preventParticleColoring method. This can be done dynamically based on block's state and position. You can take [mod I am working on](https://github.com/Nullcaller/Acro) as an example for now, the thing is done in nc.acro.util.materials.MaterialOre, but I shall move this functional to some library mod. The mod has a system that allows you to create materials "in one click", and I needed it to prevent ore particles from coloring, because it looks terrible with colored particles. You can look for realization of this "in-one-click" materials in nc.acro.mod.lists.Materials.

### CODE REVIEW

**Comment by Alpvax** [comment]
Should at least be under a client, rendering package, and should probably also be @SideOnly(Client).

File: patches/minecraft/net/minecraft/client/particle/ParticleDigging.java.patch [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ import net.minecraft.world.World;
++import net.minecraftforge.common.IParticleColorControlling;
+ import net.minecraftforge.fml.relauncher.Side;
+ import net.minecraftforge.fml.relauncher.SideOnly;
+ 
+@@ -36,8 +38,9 @@
+     public ParticleDigging setBlockPos(BlockPos pos)
+     {
+         this.sourcePos = pos;
+-
```
Comment (Alpvax) [inline patches/minecraft/net/minecraft/client/particle/ParticleDigging.java.patch]: Standard approach for patches is to leave whitespace as-is

**Comment by williewillus** [comment]
to be honest, I'd rather wait for 1.12 with java 8 and add a default method to the IBlockColor interface

**Comment by Nullcaller** [comment]
I'll do it in an hour of time.

**Comment by Nullcaller** [comment]
There. I fixed the whitespaces and redone everything with IBlockColor. Actually, it looks better to me, thanks.

**Comment by Choonster** [comment]
Assuming that `IBlockColor` is implemented on the `Block` is a bad idea. Vanilla doesn't require or do this itself, it uses completely separate (anonymous) classes to implement `IBlockColor`. I imagine a lot of mods do too.

If you do this with a method in `IBlockColor`, you'll likely need to add a method in `BlockColors` to look up the `Block`'s `IBlockColor`, call the new method and return the result (like `BlockColors#getColor`/`BlockColors#colorMultiplier` do).

**Comment by Nullcaller** [comment]
Something like that, I suppose. I don't have vanilla source to check if the last patch for BlockColors is working, but I think it will. I changed something manually, but it is quite messy. The patch does nothing with methods' names, but in forgeSrc they are different, so if I try to reverse the patch on the file from forgeSrc, it failes.

**Comment by LexManos** [comment]
Why did you randomly decide to stick it on a random vanilla interface? Making this patch 10x more complicated and causing errors in existing code?

**Comment by Nullcaller** [comment]
IBlockColor wasn't random. Blocks implementing this interface are the only blocks that need the feature. I can revert it back to version with IParticleColorControlling right now.

**Comment by Nullcaller** [comment]
Here.

**Comment by LexManos** [comment]
My point is this is only ever implemented on a Block, why not just PUT IT ON BLOCK!
Also, why is this for 1.10.2....

File: patches/minecraft/net/minecraft/client/particle/ParticleDigging.java.patch [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+ 
+ import javax.annotation.Nullable;
+ import net.minecraft.block.Block;
+ import net.minecraft.block.state.IBlockState;
+ import net.minecraft.client.Minecraft;
+@@ -9,6 +10,7 @@
+ import net.minecraft.init.Blocks;
+ import net.minecraft.util.math.BlockPos;
+ import net.minecraft.world.World;
++import net.minecraftforge.common.IParticleColoringControlling;
```
Comment (diesieben07) [inline patches/minecraft/net/minecraft/client/particle/ParticleDigging.java.patch]: Don't change imports in patch files.

**Comment by Nullcaller** [comment]
Okay, I am really dumb) Thanks, @LexManos. This is for 1.10.2 because, well, I needed it for 1.10.2, so I've done it for 1.10.2. It is done differently for 1.11.2, as far as I can see.

**Comment by diesieben07** [comment]
I don't think new features are accepted into 1.10.2.

**Comment by Nullcaller** [comment]
Okay. I shall do it for 1.11.2.

**Comment by KnightMiner** [comment]
Couldn't you already do this simply by returning no tint on tint index 0 and starting your mods tints with index 1? Particles color using a tint index of 0 specifically.

**Pull request closed without merging**